src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp

```diff
@@ -0,0 +1,127 @@
+/**
+ * @file methods/xgboost/loss_functions/sse_loss.hpp
+ * @author Rishabh Garg
+ *
+ * The sum of squared error loss class, which is a loss funtion for gradient
+ * xgboost based decision trees.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_XGBOOST_LOSS_FUNCTIONS_SSE_LOSS_HPP
+#define MLPACK_METHODS_XGBOOST_LOSS_FUNCTIONS_SSE_LOSS_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace ensemble {
+
+/**
+ * The SSE (Sum of Squared Errors) loss is a loss function to measure the
+ * quality of prediction of response values present in the node of each
+ * xgboost tree. It is also a good measure to compare the spread of two
+ * distributions. We will try to minimize this value while training.
+ *
+ * Loss = 1 / 2 * (Observed - Predicted)^2
+ */
+class SSELoss
+{
+ public:
+  /**
+   * Returns the initial predition for gradient boosting.
+   */
+  template<typename VecType>
+  typename VecType::elem_type InitialPrediction(const VecType& values)
+  {
+    return arma::accu(values) / (typename VecType::elem_type) values.n_elem;
+  }
+
+  /**
+   * Returns the first order gradient of the loss function with respect to the
+   * values.
+   *
+   * This is primarily used in calculating the residuals and split gain for the
+   * gradient boosted trees.
+   *
+   * @tparam T The type of input data. This can be both a vector or a scalar.
+   * @param observed The true observed values.
+   * @param values The values with respect to which the gradient will be
+   *     calculated.
+   */
+  template<typename T>
+  T Gradients(const T& observed, const T& values)
+  {
+    return - (observed - values);
+  }
+
+  /**
+   * Returns the second order gradient of the loss function with respect to the
+   * values. This is used only for scalars.
+   */
+  template<typename T>
+  T Hessians(const T& /* observed */, const T& /* values */)
+  {
+    return (T) 1;
+  }
+
+  /**
+   * Returns the second order gradient of the loss function with respect to the
+   * values. This is used only for vectors.
+   */
+  template<typename VecType,
+           class = std::enable_if_t<arma::is_Col<VecType>::value ||
+               arma::is_Row<VecType>::value>>
```

Comment:
I *think* (but am not sure) that you can use `VecType::is_row || VecType::is_col` here as the condition. :+1:

Reply:
I will give it a shot. If it works, then it will be nice. :)